### DISCUSSION TITLE
Training time

### DISCUSSION THREAD

**Discussion by asudeeaydin** [fangwei123456/SpikingFlow#244] [Q&A] [answered]
Hi, I'm training a network using your framework and the training time is extremely slow. I'm wondering whether there is a problem on my side or if this is expected. 

Here are the details.
My network has more or less 10 convolutional layers followed by a LIF activation layer and a batch normalization layer.
The input to the network is of dimension (8, 2, 256, 256). 
I'm using the step-by-step mode, so in each training sample, I have 10 timestep bins which I provide to the network in a for loop. 
I'm training with BPTT without truncation. 
I have 81875 samples in my training set. 

The training for a single epoch takes slightly more than 6 hours. 

In comparison, a conventional CNN takes less than an hour for a network of similar architecture with 20 channels (channels*timebins), with standard BP, same resolution, same dataset size, and same batch size. 

The difference seems too drastic so I'm wondering if this is expected, or is it an issue from my side? 

Additionally, when I analyzed the time with the Pytorch profiler I saw the following: 
The average time duration for a batch of size 8 is 2,260,800 us.

The 3 most time consuming operations were as follows:

1) aten::cudnn_convolution_backward_weight at 882,443 us
2) aten::cudnn_convolution_backward_input at 485,788 us
3) aten::cudnn_convolution at 362,311 us

From these results I'm seeing most of the computation time is spent during BP, so it can't be a faulty data loader implementation.

**Comment by asudeeaydin** [comment]
I've also tried implementing the layer-by-layer mode as it should be quicker, however since the GPU utilization is extremely high for both cases it didn't show an improvement. 

Additionally, I'm using the spikingjelly = 0.0.0.0.6 version.

**Comment by fangwei123456** [comment]
Hi, can you show me the minimal codes of the network defining, data loading and training loop?

**Comment by asudeeaydin** [comment, marked as answer]
Sorry for the delayed response. I discovered the slow running time was mainly due to the high resolution of the dataset.
